tests/test_managed_group.rs

```diff
@@ -0,0 +1,341 @@
+    );
+}
+/// Event listener function for application messages
+/// `(managed_group: &ManagedGroup, aad: &[u8], sender: &Sender, message:
+/// &[u8])`
+fn app_message_received(
+    managed_group: &ManagedGroup,
+    _aad: &[u8],
+    sender: &Sender,
+    message: &[u8],
+) {
+    println!(
+        "Message received in group {} by {} from {}: {}",
+        str::from_utf8(&managed_group.group_id().as_slice()).unwrap(),
+        str::from_utf8(&managed_group.client_id()).unwrap(),
+        str::from_utf8(managed_group.member(sender.as_leaf_index()).get_identity()).unwrap(),
+        str::from_utf8(message).unwrap()
+    );
+}
+/// Event listener function for invalid messages
+/// `(managed_group: &ManagedGroup, aad_option: Option<&[u8]>, sender_option:
+/// Option<&Sender>, error: InvalidMessageError)`
+fn invalid_message_received(
+    managed_group: &ManagedGroup,
+    aad_option: Option<&[u8]>,
+    sender_option: Option<&Sender>,
+    error: InvalidMessageError,
+) {
+    println!(
+        "Invalid message received in group {} by {}, error: {}",
+        str::from_utf8(&managed_group.group_id().as_slice()).unwrap(),
+        str::from_utf8(&managed_group.client_id()).unwrap(),
+        error,
+    );
+    if let Some(aad) = aad_option {
+        println!(" >>> AAD: {:?}", aad);
+    }
+    if let Some(sender) = sender_option {
+        println!(
+            " >>> Sender: {:?}",
+            managed_group.member(sender.as_leaf_index())
+        );
+    }
+}
+
+/// This test simulates various group operations like Add, Update, Remove in a
+/// small group
+///  - Alice creates a group
+///  - Alice adds Bob
+///  - Alice sends a message to Bob
+///  - Bob updates and commits
+///  - Alice updates and commits
+///  - Bob adds Charlie
+///  - Charlie sends a message to the group
+///  - Charlie updates and commits
+///  - Charlie removes Bob
+#[test]
+#[should_panic]
+fn managed_group_operations() {
+    for ciphersuite in Config::supported_ciphersuites() {
+        let group_id = GroupId::from_slice(b"Test group");
+
+        // Define credential bundles
+        let alice_credential_bundle =
+            CredentialBundle::new("Alice".into(), CredentialType::Basic, ciphersuite.name())
+                .unwrap();
+        let bob_credential_bundle =
+            CredentialBundle::new("Bob".into(), CredentialType::Basic, ciphersuite.name()).unwrap();
+
+        // Generate KeyPackages
+        let alice_key_package_bundle =
+            KeyPackageBundle::new(&[ciphersuite.name()], &alice_credential_bundle, vec![]).unwrap();
+
+        let bob_key_package_bundle =
+            KeyPackageBundle::new(&[ciphersuite.name()], &bob_credential_bundle, vec![]).unwrap();
+        let bob_key_package = bob_key_package_bundle.get_key_package().clone();
+
+        // Define the managed group configuration
+
+        let update_policy = UpdatePolicy::default();
+        let callbacks = ManagedGroupCallbacks::new(
+            Some(validate_add),
```

Comment:
We should move all these functions to a util module. We'll be re-using them across tests I guess.